### DISCUSSION TITLE
Does it support multi-server mode to provide better performance and stability.

### DISCUSSION THREAD

**Discussion by bandaot** [kriszyp/lmdb-js#100] [Q&A]
Thanks for providing this fancy lib for the node environment, but I'm not sure if I can run it on architecture that has multiple servers with NFS ( or AWS EFS exactly ) to share the same DB file. I can limit only one server can write to the file and NFS/EFS can provide the file lock. But I suppose the modification on file may not notify other servers to reload the dirty pages. Does anyone know if there is a way for it?

**Comment by kriszyp** [comment]
I don't know the answer to this; as I understand it, using memory maps on network files can vary quite a lot between different OS/implementations, and I am not sure of AWS EFS (although I would assume it is a pretty good implementation), and perhaps it is worth testing. I will mention that the cache coherency requirement for stability/integrity of readers (assuming you are writing on just one server) would primarily just be consistent ordering of cache updates (if there is a write to page 1 and then to page 2, the change to page 2 shouldn't be visible before the change to page 1). Anyway, love to hear if anyone has any experience with this, or if you find out anything from testing.

**Reply by OzySky** [reply]
I tried running lmdb on an EFS mount, with 1 writer container and multiple reader containers (distributed).
I first tried running the reads with ```readOnly``` and they crashed with cryptic error ```permission denied``` (also hacked the lib to pass the noLock flag to liblmdb with no effect).
Then without the readOnly flag, some readers seem to work, but some crash with error (seemingly without relation to host):
```
Error: Resource temporarily unavailable
    at renewReadTxn (file:///usr/src/node_modules/lmdb/read.js:466:18)
    at Object.ensureReadTxn (file:///usr/src/node_modules/lmdb/read.js:109:5)
```
What do you think?

**Reply by kriszyp** [reply]
With the latter error, is it possible that it is reaching some limit on the number of hosts that are acquiring a lock? Or is it pretty random when that occurs? If it is indeed "temporary", I could try implementing a retry mechanism on renewing transactions, perhaps that would be helpful?

**Reply by OzySky** [reply]
Ok, perhaps that will solve it

**Reply by kriszyp** [reply]
v2.1.3 should have improved handling of read-only stores and a retry mechanism for starting read txns, let me know if that makes any difference.
